server/parsec/components/memory/datamodel.py

```diff
@@ -249,10 +253,149 @@ class MemoryInvitation:
+    greeting_sessions: dict[UserID, MemoryGreetingSession] = field(default_factory=dict)
+
     @property
     def is_deleted(self) -> bool:
         return self.deleted_on is not None
 
+    @property
+    def is_completed(self) -> bool:
+        return self.deleted_reason == MemoryInvitationDeletedReason.FINISHED
+
+    @property
+    def is_cancelled(self) -> bool:
+        return self.deleted_reason == MemoryInvitationDeletedReason.CANCELLED
+
+    def get_greeting_session(self, user_id: UserID) -> MemoryGreetingSession:
+        try:
+            return self.greeting_sessions[user_id]
+        except KeyError:
+            greeting_session = MemoryGreetingSession(
+                token=self.token,
+                greeter_id=user_id,
+            )
+            return self.greeting_sessions.setdefault(user_id, greeting_session)
+
+
+@dataclass(slots=True)
+class MemoryGreetingSession:
+    # Immutable properties
+    token: InvitationToken
+    greeter_id: UserID
+
+    # Mutable properties
+    greeting_attempts: list[GreetingAttemptID] = field(default_factory=list)
+
+    def get_active_greeting_attempt(self, org: MemoryOrganization) -> MemoryGreetingAttempt:
+        for attempt_id in self.greeting_attempts:
+            attempt = org.greeting_attempts[attempt_id]
+            if attempt.is_active():
+                return attempt
+        attempt = MemoryGreetingAttempt(
+            greeting_attempt=GreetingAttemptID.new(),
+            token=self.token,
+            greeter_id=self.greeter_id,
+        )
+        org.greeting_attempts[attempt.greeting_attempt] = attempt
+        self.greeting_attempts.append(attempt.greeting_attempt)
+        return attempt
+
+    def new_attempt_for_greeter(
+        self, org: MemoryOrganization, now: DateTime
+    ) -> MemoryGreetingAttempt:
+        while True:
+            current_attempt = self.get_active_greeting_attempt(org)
+            current_attempt.greeter_join_or_cancel(now)
+            if current_attempt.is_active():
+                return current_attempt
+
+    def new_attempt_for_claimer(
+        self, org: MemoryOrganization, now: DateTime
+    ) -> MemoryGreetingAttempt:
+        while True:
+            current_attempt = self.get_active_greeting_attempt(org)
+            current_attempt.claimer_join_or_cancel(now)
+            if current_attempt.is_active():
+                return current_attempt
+
+
+@dataclass(slots=True)
+class MemoryGreetingAttempt:
+    # Immutable properties
+    greeting_attempt: GreetingAttemptID
+    token: InvitationToken
+    greeter_id: UserID
+
+    # Mutable properties
+    claimer_joined: None | DateTime = None
+    greeter_joined: None | DateTime = None
+    cancelled_reason: None | tuple[GreeterOrClaimer, CancelledGreetingAttemptReason, DateTime] = (
+        None
+    )
+    greeter_steps: list[bytes] = field(default_factory=list)
+    claimer_steps: list[bytes] = field(default_factory=list)
+
+    class StepOutcome(Enum):
```

Comment:
Why is stepOutcome defined inside of MemoryGreetingAttempt ?

Could it be shared with the postgre impl ?

Reply:
Because it's a return status for two methods of `MemoryGreetingAttempt`. This allows stuff like this without an extra import:
```python
        match attempt.greeter_step(step_index, greeter_data):
            case attempt.StepOutcome.MISMATCH:
                [...]
```
It's a matter of taste though, and I don't think it's worth sharing with the postgre impl.

Reply:
Ok got it ^